### ISSUE TITLE
no-extraneous-dependencies should not report the package itself

### ISSUE THREAD

**Issue by dmail** [un-ts/eslint-plugin-import-x#305]
The following `package.json`

```json
{
  "name": "package-name"
}
```

And `index.js`:

```js
import { foo } from "package-name";
```

**Actual**

`import-x/no-extraneous-dependencies` rule reports an error with the following message.

```console
'package-name' should be listed in the project's dependencies. Run 'npm i -S package-name' to add it
```

**Expected**

`import-x/no-extraneous-dependencies` remains silent because the package can import itself.

**Comment by JounQin** [comment]
Is importing the package itself a good practice?

**Comment by dmail** [comment]
My goal with this rule is mainly to ensure my dependencies are properly listed in the package.json.

Ideally I could also use it to ensure my devDependencies are listed. So I would like to have an option like allowSelfImport that I would set to true only in test files. Is that something that make sense?

**Comment by JounQin** [comment]
> But ideally I could also use it to ensure my devDependencies are listed. So I would like an option like allowSelfImport that I would set to true only in test files. Is that something that make sense?

That could make sense, but I need to think about it more and discussion with others.

cc @Shinigami92 @SukkaW @thepassle

**Comment by Shinigami92** [comment]
Can't you add it as internal devDependency?

Like I did in https://github.com/salsita/node-pg-migrate/blob/main/package.json#L114

**Comment by dmail** [comment]
> Can't you add it as internal devDependency?

That would likely fix my issue indeed 🤔 I'll try this

(By the way I think `file:.` dependencies can be declared with `./`, at least that works for me)

**Comment by JounQin** [comment]
> (By the way I think `file:.` dependencies can be declared with `./`, at least that works for me)

By the way, I'd prefer `link:.` for `yarn`. 😁

**Comment by dmail** [comment]
In the end I've added the package name into my devDependencies as follow:

```json
{
  "name": "package-name",
  "devDependencies": {
    "package-name": "./"
  }
}
```

For me it's a good enough way to be able to reference the package from test files while keeping `import-x/no-extraneous-dependencies` rules enabled.

Thanks to the quick responses and feedback.

**Comment by JounQin** [comment]
Can this issue be closed then?

**Comment by remcohaszing** [comment]
Can this be reopened? I like this rule, but it’s also the one I have to disable very often because of this issue. It looks easy to fix too. I’ll gladly provide a PR if this is accepted.

> Is importing the package itself a good practice? Isn't it a circle dependency?

This was already brought up, but I think it’s a great practice to import a package using its name in test files. This tests whether the package exports are correct.

> I believe you should disable `no-extraneous-dependencies` rule for test files? The source and test should have different rules, and this plugin cannot determine that automatically.

I agree they should have different rules. For tests I allow `devDependencies` in addition to the other kinds of dependencies. This rule is still useful for tests.

> Can't you add it as internal devDependency?

That might work, but it would only be a workaround for this issue.

**Comment by JounQin** [comment]
Hi, @remcohaszing A PR resolving this issue would be great for sure! ❤️

**Comment by neodescis** [comment]
Another use case where a package can refer to itself is a file in one package entry point importing from another, e.g. a secondary entry point referencing a primary. One of my Angular library projects uses this pattern quite a bit. Interestingly, the original eslint-plugin-import works fine in this case, but this plugin does not.

**Comment by neodescis** [comment]
@JounQin, can the PR from @remcohaszing be merged and published? I didn't see anything still holding it up, but I could be wrong. If not, I'll have to continue using the original plugin for this rule.

**Comment by neodescis** [comment]
I'm setting up lint for another project, and I'm still having to use the original eslint-plugin-import because of this. @SukkaW or @JounQin, can we get the fix merged?

**Comment by neodescis** [comment]
The irony of this not getting addressed is that we're stuck with the original eslint-plugin-import package as a result, which means we're stuck on eslint 9. Isn't the lack of support on that package the reason this one was created? What would it take to get this edge case functional in this package?

**Comment by JounQin** [comment]
> The irony of this not getting addressed is that we're stuck with the original eslint-plugin-import package as a result, which means we're stuck on eslint 9. Isn't the lack of support on that package the reason this one was created? What would it take to get this edge case functional in this package?

@neodescis Did you mean the original `import` plugin doesn't have this problem while we did? Or something else?
